patch/patch.go

```diff
@@ -87,3 +91,81 @@ func WalkPatches(dir string, fn func(string) error) error {
 	}
 	return nil
 }
+
+// FindAncestorConfig finds and reads the config file governing dir. Searches dir and all ancestor
+// directories of dir, similar to how ".git" files work.
+//
+// If no config file is found in any ancestor, and an ancestor dir appears to be like microsoft/go
+// by convention (contains a "patches" directory and a "go" directory), creates a config struct to
+// fit the conventional repo.
+func FindAncestorConfig(dir string) (*FoundConfig, error) {
+	originalDir := dir
+	var byConvention *FoundConfig
+	for {
+		c, err := findConfigInDir(dir)
+		if err != nil {
+			return nil, err
+		}
+		if c != nil {
+			// Found the config file in dir.
+			return c, nil
+		}
+		// We didn't find a config file, but check if we found a conventional fork directory. Keep
+		// track of only the first one (most nested one) that we find.
+		if byConvention == nil {
+			byConvention, err = dirConfig(dir)
+			if err != nil && !errors.Is(err, os.ErrNotExist) {
+				log.Printf("Unable to determine if %v is a Go directory for a surprising reason: %v", dir, err)
+			}
+		}
+
+		// We didn't find the config file yet. Find the parent dir for the next iteration.
+		parent := filepath.Dir(dir)
+		// When we've hit the filesystem root, Dir goes no further.
+		if dir == parent {
+			if byConvention != nil {
+				return byConvention, nil
+			}
+			return nil, fmt.Errorf("no %q file or Microsoft Go root found in any ancestor of %v", ConfigFileName, originalDir)
+		}
+		dir = parent
+	}
+}
+
+func dirConfig(dir string) (*FoundConfig, error) {
+	if ok, err := isDir(filepath.Join(dir, conventionalConfig.SubmoduleDir)); !ok {
+		return nil, err
+	}
+	if ok, err := isDir(filepath.Join(dir, conventionalConfig.PatchesDir)); !ok {
+		return nil, err
+	}
+	return &FoundConfig{
+		Config:  conventionalConfig,
+		RootDir: dir,
+	}, nil
+}
+
+func isDir(dir string) (ok bool, err error) {
+	if info, err := os.Stat(dir); err != nil {
+		return info != nil && info.IsDir(), err
+	}
+	return true, nil
```

Comment:
Won't passing a normal file path, not a directory, end up hitting this line?

Reply:
Yeah, looks like I put the `info != nil && info.IsDir()` in the wrong place. Will fix it up. (I think naming the variable `dir` even though we don't know it's a dir yet makes it confusing, so I'll change that too.)